### PR TITLE
Feat/ml integeration test

### PR DESCRIPTION
**Pull request by db091756**
### PR Type

- Feature
- Tests

### Description

Adds an example showcasing using a coreset in a ML application.

### How Has This Been Tested?

Corresponding test added,

### Does this PR introduce a breaking change?

(Write your answer here.)

### Screenshots

(Write your answer here.)

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: examples/coreset_in_model_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    random_fit_time = time()
+    random_model = KernelRidge(
+        kernel="rbf", gamma=float(median_heuristic(x[random_indices])), alpha=1e-1
+    ).fit(x[random_indices], y[random_indices])
+    random_fit_time = time() - random_fit_time
+    random_mse = jnp.linalg.norm(y_test - random_model.predict(x_test))
+
+    # Produce some scatter plots
+    x_plot = jnp.linspace(x.min(), x.max(), 100).reshape(-1, 1)
+    plt.scatter(x, y, s=20, color="black", alpha=0.25, label="Data")
```
Comment (rg936672) [inline examples/coreset_in_model_pipeline.py:144]: I think it's worth adding a comment before each plot and a line break between them, so there's a bit more visual separation between the code for the different plots.

**Review by rg936672 — CHANGES_REQUESTED** [review]
A couple of minor points, but by and large this looks great.

File: examples/coreset_in_model_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+    # Setup the original data object
+    data = Data(x)
+
+    # Set the bandwidth parameter of the kernel using a median heuristic derived from at
+    # most 1000 random samples in the data.
+    num_samples_length_scale = min(num_data_pairs, 1_000)
+    generator = np.random.default_rng(seed)
+    idx = generator.choice(num_data_pairs, num_samples_length_scale, replace=False)
+    length_scale = median_heuristic(x[idx]).item()
```
[the remark this answers is not in the document]
Comment (db091756) [reply examples/coreset_in_model_pipeline.py:102]: The median heuristic is used not only to build the coreset, but also to fit each kernel ridge regression model (full data, coreset and random). I could add this time to all the models but I think its probably fine to leave this as is.

File: examples/coreset_in_model_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+    if out_path is not None:
+        if not out_path.is_absolute():
+            out_path = Path(__file__).parent.joinpath(out_path)
+        plt.savefig(out_path)
+
+    return (
+        (float(full_data_mse), float(full_data_fit_time)),
+        (float(coreset_mse), float(coreset_overall_time)),
+        (float(random_mse), float(random_fit_time)),
```
[the remark this answers is not in the document]
Comment (db091756) [reply examples/coreset_in_model_pipeline.py:234]: Added the timings to the final graph, does this suffice?

File: examples/coreset_in_model_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    length_scale = median_heuristic(x[idx]).item()
+    kernel = SquaredExponentialKernel(length_scale=length_scale)
+
+    print("Estimating kernel ridge regression model with full data...")
+    full_data_fit_time = time()
+    full_data_model = KernelRidge(kernel="rbf", gamma=length_scale, alpha=1e-2).fit(
+        x, y
+    )
+    full_data_fit_time = time() - full_data_fit_time
+    full_data_mse = jnp.linalg.norm(y_test - full_data_model.predict(x_test))
```
Comment (rg936672) [inline examples/coreset_in_model_pipeline.py:112]: I missed this the first time round, but this isn't the mean squared error, it's the square root of the sum of squared errors. (This is 1.875 on my machine, which is clearly incorrect looking at the graph!) The same issue is present for the other two "MSE" variables.

**Review by rg936672 — CHANGES_REQUESTED** [review]
Looks good, one mathematical error that I missed the first time round

File: examples/coreset_in_model_pipeline.py (hunk 4 above)
Comment (rg936672) [reply examples/coreset_in_model_pipeline.py:112]: (I'd prefer that the statistic displayed on the graph was the root mean squared error, as that's more readily interpretable than the mean squared error.)

Comment (db091756) [reply examples/coreset_in_model_pipeline.py:112]: Good spot, will fix

**Review by rg936672 — APPROVED** [review]

**Pull request merged**